### PR TITLE
ARROW-693: [Java] Add dictionary support to JSON reader and writer

### PR DESCRIPTION
**Pull request by BryanCutler**
Added support for Dictionary encoded vectors for reading and writing JSON.  Also added validation to compare DictionaryEncodings and DictionaryBatches.

New unit test for round-trip of flat dictionary type.

### CODE REVIEW

**Comment by BryanCutler** [comment]
@julienledem please take a look when you can.  I did some minor refactoring of ArrowReader/Writer to allow reuse of `toMessageFormat` and `toMemoryFormat` for Fields with dictionary encodings.

One thing I couldn't quite figure out, was creating a DictionaryWriter.  I tried a simple approach but couldn't get compilation working with the codegen and related classes.  Could you maybe comment on how you see this fitting in and maybe what semantics it would have for writing both the data vector and dictionary vector?

cc @wesm, this should be good enough to start testing but I'm pretty sure it will be expecting the `typeLayout` to be the dictionary type, but I can give it a try.

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: It seems odd having all these mutable objects! =) @julienledem do you think there's a reasonable path to having a little more immutability in the internals, or is it not practical?

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Is there a JIRA about handling encoded subfields? We should tackle this sooner rather than later. As long as this doesn't leak into the user API I think it's OK

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java]: Can the dictionary-encoded fields be encapsulated in the schema? This is an additional spillover of technical debt from the encoded subfield issue

File: java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java]: This is a slight oddness. Will users be exposed to this at all?

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: Change these names to be more generic?

File: java/vector/src/test/java/org/apache/arrow/vector/file/json/TestJSONFile.java
Comment (wesm) [inline java/vector/src/test/java/org/apache/arrow/vector/file/json/TestJSONFile.java]: Can you expand this test to include

* More than 1 dictionary
* Two vectors using the same dictionary

**Review by wesm — COMMENTED** [review]
thanks @BryanCutler for your work on this!

@julienledem I'll let you take a closer look. The dictionary stuff in Java is going to require some work to support nested subfields (out of scope for this patch). I see dictionary-encoded fields as being an important (and likely very common) use case -- we could have cases where dictionaries are expanding as new data is read into memory

**Comment by elahrvivaz** [comment]
@wesm I'll take a look at this tomorrow, thanks

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java]: I'm not sure how big of change that would be, but I think it would then tie the schema to an allocator which might cause problems.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java]: I agree it is a little odd. This is for internal Arrow use only, so users shouldn't be exposed to it.  The reason for this (please correct me if I'm wrong @julienledem) is so the encoded field in a record batch will have the dictionary type and before that field is loaded, the type is switched to be the index type to match the actual data in the batch so it can be read as normal.  I'm not sure if it's possible to infer the dictionary type (I think you mentioned this is what is done in C++), but I can look into that.

File: java/vector/src/test/java/org/apache/arrow/vector/file/json/TestJSONFile.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/file/json/TestJSONFile.java]: Yeah, I definitely want to do that.  I was just a little pressed for time and wanted to put what I had so far out.

**Comment by BryanCutler** [comment]
Thanks for the review @wesm!  I have been a little short on time, but I will hopefully have some time in the next couple days to do some testing and look into the comments above.

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java]: aren't the dictionary-encoded fields already encapsulated in the schema? Or do you mean the dictionary batches?

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: I believe subfields can be dictionary encoded currently. We use a map vector with dictionary encoded children just fine.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java]: users aren't exposed to it - it would be nice to have to do it, but it would require integrating dictionary-awareness throughout the java vector code base, which would be a fairly large change. the initial dictionary work was basically bolted on top of the existing code to minimize changes.

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: we should probably avoid sorting the input collection - the user might not be expecting that. instead, create a new collection and sort that.

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: this comment is confusing... isn't the order important, which is why we're sorting? or are you just indicating the actual order doesn't matter as long as the two lists are aligned?

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: if most of these methods haven't changed, and only formatting was fixed, it might be good to revert them to avoid confusion. not sure what the project standards are in that regards...

**Review by elahrvivaz — COMMENTED** [review]
this looks good to me, thanks

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileWriter.java]: Yes, the dictionary-encoded fields are in the schema.  This line is creating a new schema with the same dictionary-encoded fields having switched the `typeLayout` to be the dictionary type.

File: java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java]: I removed it because the check is already done [here](https://github.com/apache/arrow/pull/752/files/8a323cf1d190a2133b620b980af2aaeebcff4ef8#diff-c47bcf8f4cc51762974cc2968da823f9R71) and the semantics are slightly different now:

1) Goes through each field in the schema
2) If field has an encoding, then verify the lookup dictionary in the provider (with the null check)
3)  Add the id to a set of dictionary ids that are being used by Fields
4) For each dictionary id being used, write the dictionary record batch (no need to check if this id is in the provider at this point because it was already done)

Hopefully that sounds ok

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: Yeah, most of this was formatting changes.. not sure if this was not the right time to fix it.  I can revert if this is too much to change?

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: I think what I was going for here was that you could compare dictionaries from 2 files even if the schema was different.  Looking at this again, that doesn't seem like it would even work.  I'll remove the sorting which will be fine because this is only used where the 2 schemas are equal.

**Comment by BryanCutler** [comment]
Thank you for the review @elahrvivaz !  I'll update soon

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]:
```
+ " expected end of array at " + parser.getTokenLocation()
```

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java]: shouldn't the property be just "ordered" ?

File: java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java]: Yes, it sounds like we should do something better but I have not looked at that code deep enough recently to suggest a better approach. Since this is not introduced by this change this could be fixed in a subsequent change.

File: java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/util/Validator.java]: add arrowVectors and jsonVectors to the message? This would help with understanding why the sizes are different.

**Review by julienledem — APPROVED** [review]
I made some comments.
Overall this looks good to me.

File: java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java]: :+1:

**Comment by wesm** [comment]
I'll write the actual integration tests this week so we can get those working. We should push for a 0.5.0 release after that

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: Yeah, I agree on throwing an exception.  This was taken from ArrowReader, I'll make the change there also to be consistent unless you object.

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java]: That would work a little easier with the annotations, so doesn't make a difference to me.  I was following the naming from the DictionaryEncoding table in schema.fbs - @wesm  is "ordered" is fine with you also?

**Comment by BryanCutler** [comment]
Sorry for the delay, I ran into some trouble trying to clean up the tests.  I expanded the flat dictionary test to include a second dictionary and 2 fields accessing the same dictionary.  Also added a test for nested field with dictionary in JSON and seems to be working fine.

**Comment by BryanCutler** [comment]
Please take another look when you can @julienledem , I think it's good to go now.

File: java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryProvider.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/DictionaryProvider.java]: I added this to be able to iterate over all dictionaries for testing

File: java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/file/json/JsonFileReader.java]: I'm quite surprised. At least, we have no unit tests to verify this one way or the other. Opened https://issues.apache.org/jira/browse/ARROW-1175

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/DictionaryEncoding.java]: If it's no problem, `isOrdered` is more consistent with Schema.fbs

File: java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java
Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java]: I think it's OK if in the memory format, the fields always have the index type (and you have to rely on having the dictionaries somewhere to do any analytics). Having to alter the schema before serializing to a message may best be avoided

**Review by wesm — APPROVED** [review]
This LGTM, I need to rebase my data generator patch on this and see if the integration tests pass. I will report back

**Comment by wesm** [comment]
@BryanCutler I updated #800 to help with reconciling discrepancies. I found a couple problems

* The Java JSON reader is expecting the name of the dictionary in the embedded batch to have the form `DICT$ID`, e.g. `DICT5`. This seems brittle since this name is never used -- at a glance this didn't look easy to fix so I'm generating that name in the integration test generator

* The C++ implementation does not preserve the dictionary ids in the schema; they are generated during the IPC process 

From where I'm at now:

```
17:20:16.942 [main] DEBUG org.apache.arrow.tools.Integration - Arrow Input file size: 3650
17:20:16.943 [main] DEBUG org.apache.arrow.tools.Integration - ARROW schema: Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 0], dict2_0: Int(16, true)[dictionary: 1]>
17:20:16.943 [main] DEBUG org.apache.arrow.tools.Integration - JSON Input file size: 7896
17:20:16.943 [main] DEBUG org.apache.arrow.tools.Integration - JSON schema: Schema<dict1_0: Int(8, true)[dictionary: 0], dict1_1: Int(32, true)[dictionary: 0], dict2_0: Int(16, true)[dictionary: 1]>
Incompatible files
Different dictionaries:
Dictionary DictionaryEncoding[id=0,ordered=false,indexType=Int(8, true)] org.apache.arrow.vector.NullableVarCharVector@245b4bdc[name = DICT0, ...]
Dictionary DictionaryEncoding[id=0,ordered=false,indexType=Int(8, true)] org.apache.arrow.vector.NullableVarCharVector@6c64cb25[name = DICT0, ...]
17:20:16.946 [main] ERROR org.apache.arrow.tools.Integration - Incompatible files
java.lang.IllegalArgumentException: Different dictionaries:
Dictionary DictionaryEncoding[id=0,ordered=false,indexType=Int(8, true)] org.apache.arrow.vector.NullableVarCharVector@245b4bdc[name = DICT0, ...]
Dictionary DictionaryEncoding[id=0,ordered=false,indexType=Int(8, true)] org.apache.arrow.vector.NullableVarCharVector@6c64cb25[name = DICT0, ...]
	at org.apache.arrow.vector.util.Validator.compareDictionaries(Validator.java:75) ~[arrow-tools-0.5.0-SNAPSHOT-jar-with-dependencies.jar:na]
	at org.apache.arrow.tools.Integration$Command$3.execute(Integration.java:197) ~[arrow-tools-0.5.0-SNAPSHOT-jar-with-dependencies.jar:na]
	at org.apache.arrow.tools.Integration.run(Integration.java:114) ~[arrow-tools-0.5.0-SNAPSHOT-jar-with-dependencies.jar:na]
	at org.apache.arrow.tools.Integration.main(Integration.java:65) ~[arrow-tools-0.5.0-SNAPSHOT-jar-with-dependencies.jar:na]
```

I am going to merge the updated #800 once the build passes and then we can work on getting the integration tests to pass from this branch.

**Comment by BryanCutler** [comment]
Thanks @wesm , I'm going to be looking in to this today.

> The Java JSON reader is expecting the name of the dictionary in the embedded batch to have the form DICT$ID

Yeah, I don't like this either it is too brittle.  It was because the logic to read the dictionary batch is the same to read regular batches and includes a check that the column in the batch matches the schema, but the dictionary field gets assigned an arbitrary name when reading the schema.  I could put a flag to not check, but that would end up in a public api for the json reader.  I'll take another look and see if there is a better way to avoid this.

**Comment by wesm** [comment]
Thanks. I just merged #800 so if you rebase then you can enable the dictionary integration test. If there's any changes needed in C++ or the integration tests themselves I can help with commits for you to cherry-pick into this branch

**Comment by BryanCutler** [comment]
@wesm , I removed the need for a specific name of the dictionary, so now any name is fine.  There was a flag that is set after the dictionaries are read, so it's safe to not check the expected names until this is set.  I'm looking at the integration tests now.

**Comment by BryanCutler** [comment]
Hi @wesm , I've been running integration tests with just the Java tester.  I did make the change below to the script because the Field name was coming out empty, does that look ok?
```python
--- a/integration/integration_test.py
+++ b/integration/integration_test.py
@@ -581,6 +581,8 @@ class DictionaryType(DataType):
         return self.index_type._get_type_layout()
 
     def generate_column(self, size, name=None):
+        if name is None:
+            name = self.name
         return self.index_type.generate_range(size, 0, len(self.dictionary),
                                               name=name)
```
I have the Java side of integration tests working, I had to fix the way dictionaries were being compared which is the problem you were seeing I believe.  I'm getting the following error when I enable the CPPTester
```
/home/bryan/git/arrow/cpp/debug/debug/json-integration-test --integration --arrow=/tmp/tmp3u4minzl/674ee56ab3634a3b9dda5251de8a2995_generated_dictionary.json_to_arrow --json=/tmp/tmpzojpk0qy/generated_dictionary.json --mode=JSON_TO_ARROW
Command failed: /home/bryan/git/arrow/cpp/debug/debug/json-integration-test --integration --arrow=/tmp/tmp3u4minzl/674ee56ab3634a3b9dda5251de8a2995_generated_dictionary.json_to_arrow --json=/tmp/tmpzojpk0qy/generated_dictionary.json --mode=JSON_TO_ARROW
With output:
--------------
Found schema: dict1_0: string
dict1_1: string
dict2_0: int64
/home/bryan/git/arrow/cpp/src/arrow/ipc/json-internal.cc888 Check failed: val.IsString() 
```
Is this something you fixed on your end already?

File: java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: @elahrvivaz I removed this because when doing integration tests, the FieldVector `dictionary` wasn't being properly compared.  I think to compare Dictionaries, it's better to just separately compare the DictionaryEncoding and FieldVector, which is what I'm doing now in Integration.java.  Does that sound right to you?

**Comment by wesm** [comment]
The tests all pass on master with C++ producing and consuming

```
$ python integration_test.py 
-- C++ producing, C++ consuming
==========================================================
Testing file /home/wesm/code/arrow/integration/data/simple.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /home/wesm/code/arrow/integration/data/struct_example.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmptvdpj_7e/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmptvdpj_7e/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmptvdpj_7e/generated_datetime.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmptvdpj_7e/generated_nested.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmptvdpj_7e/generated_dictionary.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
-- All tests passed!
```

The other change you showed looks fine; I'll have to pull your latest changes and take a look

**Comment by BryanCutler** [comment]
Ah, my mistake I was pointing to an older C++ build - all tests pass for me!

File: java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: hmm, so basically `FieldVector.equals` isn't correct? I'm not sure if it would affect anything to remove this or not...

Comment (wesm) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: I am not sure why this code was wrong -- @elahrvivaz can you verify that this doesn't break your downstream application? I will hold off on merging for now

Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: I don't think that it would affect anything. IMO seems better to leave it and add a ticket to fix the vector.equals, although maybe there is a reason vectors don't implement equals.

Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: Well, this should probably be looked at as a separate issue anyway and since I'm no longer using it for this PR, I'll revert it.  I created https://issues.apache.org/jira/browse/ARROW-1184 for this.

Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java]: sounds good, thanks

**Comment by BryanCutler** [comment]
All tests pass
```
-- C++ producing, C++ consuming
==========================================================
Testing file /home/bryan/git/arrow/integration/data/struct_example.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /home/bryan/git/arrow/integration/data/simple.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_datetime.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_nested.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_dictionary.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream

-- C++ producing, Java consuming
==========================================================
Testing file /home/bryan/git/arrow/integration/data/struct_example.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /home/bryan/git/arrow/integration/data/simple.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_datetime.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_nested.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_dictionary.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream

-- Java producing, C++ consuming
==========================================================
Testing file /home/bryan/git/arrow/integration/data/struct_example.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /home/bryan/git/arrow/integration/data/simple.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_datetime.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_nested.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_dictionary.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream

-- Java producing, Java consuming
==========================================================
Testing file /home/bryan/git/arrow/integration/data/struct_example.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /home/bryan/git/arrow/integration/data/simple.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_primitive.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_datetime.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_nested.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
==========================================================
Testing file /tmp/tmpxayl0b66/generated_dictionary.json
==========================================================
-- Creating binary inputs
-- Validating file
-- Validating stream
-- All tests passed!
```

**Comment by wesm** [comment]
+1, thank you for doing this @BryanCutler! I'm glad that it seems we didn't need to make any changes to the binary format implementations, but it's good to have the peace of mind so we can refactor and add more features

**Pull request closed without merging**
